package/gluon-mesh-babel/src/respondd.c

```diff
@@ -71,13 +75,250 @@
+				route->metric = *(int *)RTA_DATA(rta);
+				if (route->metric < 0 || route->metric > KERNEL_INFINITY)
+					route->metric = KERNEL_INFINITY;
+				break;
+			default:
+				break;
+		}
+	}
+
+	return 1;
+}
+
+static bool handle_kernel_routes(const struct nlmsghdr *nh, struct kernel_route *route) {
+	int len = nh->nlmsg_len;
+	struct rtmsg *rtm;
+
+	rtm = (struct rtmsg *)NLMSG_DATA(nh);
+	len -= NLMSG_LENGTH(0);
+
+	/* Ignore cached routes, advertised by some kernels (linux 3.x). */
+	if (rtm->rtm_flags & RTM_F_CLONED) return false;
+
+	if (parse_kernel_route_rta(rtm, len, route) < 0) return false;
+
+	return true;
+}
+
+static bool rtnl_handle_msg(const struct nlmsghdr *nh,
+			    struct kernel_route *route) {
+	if (nh->nlmsg_type == RTM_NEWROUTE) {
+		handle_kernel_routes(nh, route);
+		if (!(route->plen == 0 && route->metric >= KERNEL_INFINITY))
+			return true;
+	}
+	return false;
+}
+
+static struct json_object *get_default_route(void) {
+	struct json_object *ret = json_object_new_object();
+	int nlfd = socket(AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+	if (nlfd < 0) {
+		perror("can't open RTNL socket");
+		return NULL;
+	}
+	struct sockaddr_nl snl = {
+	    .nl_family = AF_NETLINK,
+	    .nl_groups = RTMGRP_IPV6_ROUTE,
+	};
+
+	if (bind(nlfd, (struct sockaddr *)&snl, sizeof(snl)) < 0)
+		perror("can't bind RTNL socket");
+
+	struct in6_addr addr = {};
+	inet_pton(AF_INET6, "2000::/3", &addr);
+
+	get_route(nlfd, 0, &addr, 3);
+
+	uint8_t readbuffer[8192];
+	int count = recv(nlfd, readbuffer, sizeof(readbuffer), 0);
+
+	struct nlmsghdr *nh;
+	struct nlmsgerr *ne;
+	struct kernel_route route;
+
+	nh = (struct nlmsghdr *)readbuffer;
+	if (NLMSG_OK(nh, count)) {
+		switch (nh->nlmsg_type) {
+			case NLMSG_DONE:
+				break;
+			case NLMSG_ERROR:
+				perror("handling netlink error-message");
+				ne = NLMSG_DATA(nh);
+				if (ne->error <= 0)
+					break;
+				/* Falls through. */
+			default:
+				if (rtnl_handle_msg(nh, &route) == true) {
+					json_object_object_add(ret, "src", json_object_new_string(print_ip(&route.src_prefix)));
+					json_object_object_add(ret, "via", json_object_new_string(print_ip(&route.gw)));
+					char ifname[IFNAMSIZ];
+					json_object_object_add(ret, "interface", json_object_new_string(if_indextoname(route.ifindex, ifname)));
+				}
+				break;
+		}
+	}
+
+	close(nlfd);
+	return ret;
+}
+
+static int babeld_connect() {
+	int fd = -1;
+	fd_set rfds;
+	FD_ZERO(&rfds);
+
+	printf("connecting to babeld\n");
+
+	do {
+		fd = babelhelper_babel_connect(BABEL_PORT);
+		if (fd < 0) {
+			fprintf(stderr, "Connecting to babel socket failed. Retrying.\n");
+			usleep(1000000);
+		}
+	} while (fd < 0);
+
+	FD_SET(fd, &rfds);
+
+	// receive and ignore babel header
+	while (true) {
+		if ( babelhelper_input_pump(&bhelper_ctx, fd, NULL, babelhelper_discard_response))
+			break;
+
+		if (select(fd + 1, &rfds, NULL, NULL, NULL) < 0) {
+			perror("select (babel header):");
+		};
+	}
+
+	int amount = 0;
+	while (amount != 8) {
```

Comment:
Could you clarify this?

Reply:
this is the length of the message that is sent to babeld: "monitor\n"

This is subject to change as babeld is going to implement filters on the monitor interface in the future (patch pending merge upstream).

Reply:
Makes sense!